Database/MySQL.php

```diff
@@ -12,66 +12,116 @@ class MySQL extends BaseDatabase
+            foreach ($this->params['ignore_tables'] as $ignoreTable) {
+                if ($this->params['all_databases']) {
                     if (false === strpos($ignoreTable, '.')) {
                         throw new \LogicException(
                             'When dumping all databases both database and table must be specified when ignoring table'
                         );
                     }
                     $this->ignoreTables .= sprintf('--ignore-table=%s ', $ignoreTable);
                 } else {
-                    $this->ignoreTables .= sprintf('--ignore-table=%s.%s ', $this->database, $ignoreTable);
+                    $this->ignoreTables .= sprintf('--ignore-table=%s.%s ', $this->params['database'], $ignoreTable);
                 }
             }
         }
+    }
+
+    /**
+     * Prepare mysql configuration file for connection
+     */
+    protected function prepareConfigurationFile()
+    {
+        if ($this->params['db_user']) {
+            $cnfParams['user']  = $this->params['db_user'];
+            $cnfFile            = "[client]\n";
 
-        /* if user is set, we add authentification */
-        if ($params['db_user']) {
-            $this->auth = sprintf('-u%s', $params['db_user']);
+            if ($this->params['db_password']) {
+                $cnfParams = array_merge(
+                    $cnfParams,
+                    array(
+                        "password"  => $this->params['db_password'],
+                        "host"      => $this->params['db_host'],
+                        "port"      => $this->params['db_port']
+                    )
+                );
+            }
 
-            if ($params['db_password']) {
-                $this->auth = sprintf("--host=\"%s\" --port=\"%d\" --user=\"%s\" --password=\"%s\"", $params['db_host'], $params['db_port'], $params['db_user'], $params['db_password']);
+            foreach ($cnfParams as $key => $value) {
+                $cnfFile .= "$key = \"$value\"\n";
             }
+
+            $this->filesystem->dumpFile($this->getConfigurationFilePath(), $cnfFile, 0600);
+            $this->auth = sprintf("--defaults-extra-file=\"%s\" ", $this->getConfigurationFilePath());
         }
     }
 
     /**
+     * Remove mysql configuration file from backup files
+     */
+    protected function removeConfigurationFile()
+    {
+        $this->filesystem->remove($this->getConfigurationFilePath());
+    }
+
+    /**
+     * Gets mysql configuration file full path
+     * @return string
+     */
+    protected function getConfigurationFilePath()
+    {
+        return $this->dataPath . "mysql.cnf";
+    }
+
+    /**
      * {@inheritdoc}
      */
     public function dump()
     {
         $this->preparePath();
+        $this->prepareFileName();
+        $this->prepareIgnoreTables();
+        $this->prepareConfigurationFile();
         $this->execute($this->getCommand());
+        $this->removeConfigurationFile();
```

Comment:
What if an exception is thrown somewhere. I do not what to risk that a configuration file is lying around. Let's use a try/catch and make sure 'removeConfigurationFile' always is running.